pkg/kncloudevents/message_sender.go

```diff
@@ -72,3 +94,82 @@ func (s *HttpMessageSender) NewCloudEventRequestWithTarget(ctx context.Context,
 func (s *HttpMessageSender) Send(req *nethttp.Request) (*nethttp.Response, error) {
 	return s.Client.Do(req)
 }
+
+// CheckRetry specifies a policy for handling retries. It is called
+// following each request with the response and error values returned by
+// the http.Client. If CheckRetry returns false, the Client stops retrying
+// and returns the response to the caller. If CheckRetry returns an error,
+// that error value is returned in lieu of the error from the request. The
+// Client will close any response body when retrying, but if the retry is
+// aborted it is up to the CheckRetry callback to properly close any
+// response body before returning.
+type CheckRetry func(ctx context.Context, resp *nethttp.Response, err error) (bool, error)
+
+// Backoff specifies a policy for how long to wait between retries.
+// It is called after a failing request to determine the amount of time
+// that should pass before trying again.
+type Backoff func(attemptNum int, resp *nethttp.Response) time.Duration
+
+type RetriesConfig struct {
+
+	// Maximum number of retries
+	RetryMax int
+
+	CheckRetry CheckRetry
+
+	Backoff Backoff
+}
+
+func (s *HttpMessageSender) SendWithRetries(req *nethttp.Request, config RetriesConfig) (*nethttp.Response, error) {
```

Comment:
take `config` as a pointer: if `config == nil` then take the hot path to `s.Send()`